packages/nx/src/command-line/exec.ts

```diff
@@ -0,0 +1,108 @@
+import {
+  createProjectGraphAsync,
+  readProjectsConfigurationFromProjectGraph,
+} from '../project-graph/project-graph';
+import { calculateDefaultProjectName } from './run-one';
+import { workspaceRoot } from '../utils/workspace-root';
+import { readNxJson } from '../config/configuration';
+import { splitArgsIntoNxArgsAndOverrides } from '../utils/command-line-utils';
+import { ProjectGraph, ProjectGraphProjectNode } from '../config/project-graph';
+import {
+  ProjectConfiguration,
+  TargetConfiguration,
+} from '../config/workspace-json-project-json';
+import { output } from '../utils/output';
+import { exec } from 'child_process';
+import { getPackageManagerCommand } from '../utils/package-manager';
+
+export async function nxExecCommand(
+  args: Record<string, string[]>
+): Promise<void> {
+  const { overrides } = splitArgsIntoNxArgsAndOverrides(
+    args,
+    'run-one',
+    { printWarnings: false },
+    readNxJson()
+  );
+
+  const scriptSeparatorIdx = process.argv.findIndex((el) => el === '--');
+  if (scriptSeparatorIdx === -1) {
+    output.error({
+      title: 'nx exec requires passing in a command after `--`',
+    });
+  }
+
+  const scriptArgV: string[] = overrides.__overrides_unparsed__;
+
+  // NX is already running
+  if (process.env.NX_TASK_TARGET_PROJECT) {
+    return runCommand(
+      scriptArgV.reduce((cmd, arg) => cmd + `${arg} `, '').trim()
+    );
+  } else {
+    // nx exec is being ran inside of Nx's context
+    return runScriptAsNxTarget();
+  }
+}
+
+function runCommand(command) {
+  return new Promise<void>((res, rej) =>
+    exec(command).on('exit', (code) => {
+      if (code !== 0) {
+        rej(code);
+      } else {
+        res();
+      }
+    })
+  );
+}
+
+async function runScriptAsNxTarget() {
+  const projectGraph = await createProjectGraphAsync();
+  const projectName = calculateDefaultProjectName(
+    process.cwd(),
+    workspaceRoot,
+    readProjectsConfigurationFromProjectGraph(projectGraph)
+  );
+
+  if (!projectName) {
+    output.error({
+      title: 'Unable to determine project name for nx exec',
+      bodyLines: [
+        "`nx exec` should be ran from within a project's root directory.",
```

Comment:
I think this can only realistically happen when you add nx exec to an npm script of a non-project. If that's the case, I'd make the error message reflect that.